src/Bicep.Core/PrettyPrint/PrettyPrinter.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using Bicep.Core.Extensions;
+using Bicep.Core.Parser;
+using Bicep.Core.PrettyPrint.Options;
+using Bicep.Core.Syntax;
+
+namespace Bicep.Core.PrettyPrint
+{
+    public static class PrettyPrinter
+    {
+        private static readonly IReadOnlyDictionary<int, string> IndentsBySize = Enumerable.Range(1, 8)
+            .ToDictionary(size => size, size => new string(' ', size));
+
+        public static string? PrintProgram(ProgramSyntax programSyntax, PrettyPrintOptions options)
+        {
+            if (programSyntax.GetParseDiagnostics().Count > 0)
+            {
+                return null;

```

Comment:
>null [](start = 23, length = 4)

What would it take to support formatting when there are syntax errors?

Reply:
Technically the formatter will just work when there are syntax errors. However, the output is not predictable at all. Even a missing `}` could mess up a big chunk:

```bicep
param p0 object = { a: true
param p1 string
param p2 string
param p3 string
param p4 string

var foo = {
  b: false
}

```

will be formatted as

```bicep
param p0 object = {a: true
  paramp1string 
  paramp2string 
  paramp3string 
  paramp4string 

  varfoo={ 
  b: false
}

```
I think this one of the reason why formatters like Prettier and Black don't run formatting when there are lexical and syntactic errors, because there are infinite number of ways to break a syntax tree, and finetuning the code formatter to make sure it doesn't produce something that will surprise users requires a lot of efforts. 

That being said, since Bicep is not a generic programing language and is **declarative**, I do think there are certain things we can do to make it work. For example, if a declaration has syntax errors, skip formatting that declaration but keep formatting the other nodes. I'll do some experiments and send a separate PR if it goes well.

Reply:
Sounds good.

Reply:
I'm not convinced this is a problem we need to solve, especially if the perceived user experience is unpredictable (and I don't see how it wouldn't be). If there's a way to simply fail with an error if there are parse errors, that would be my preferred option.

Reply:
I disagree. I'd say sometimes doing nothing in the format command would be even worse than partially formatting a file. (I don't want the user to ever wonder why nothing happened when they clicked the format option.)

I've also never had a format command fail in an IDE with an error. Just tried in JSON, ts in VS code and they do format even with syntax errors. Same for c# in VS. 

That said, they do produce weird formatting in some cases when there are syntax errors, so we don't need to be perfect here. If the user hates the result they can always undo.

Reply:
We'll have to agree to disagree then :)
> I disagree. I'd say sometimes doing nothing in the format command would be even worse than partially formatting a file. (I don't want the user to ever wonder why nothing happened when they clicked the format option.)

100% agree we should not fail without informing the user - I'm making the assumption that we have the ability to surface an error.

> I've also never had a format command fail in an IDE with an error. Just tried in JSON, ts in VS code and they do format even with syntax errors. Same for c# in VS.

Maybe it's just me then, but that behavior seems totally undesirable...

> That said, they do produce weird formatting in some cases when there are syntax errors, so we don't need to be perfect here. If the user hates the result they can always undo.

Technically true, but there are going to be cases where it's not immediately obvious what has been changed - if you're working with larger files, without source control etc. Reformatting when we don't have matching brackets has the potential to introduce some very subtle changes that could go unnoticed for a while.

Reply:
> We'll have to agree to disagree then :)
> 
> > I disagree. I'd say sometimes doing nothing in the format command would be even worse than partially formatting a file. (I don't want the user to ever wonder why nothing happened when they clicked the format option.)
> 
> 100% agree we should not fail without informing the user - I'm making the assumption that we have the ability to surface an error.

👍 LSP protocol allows error response theoretically, but I don't know if OmniSharp can handle it. Even if we surfaced an error, VS code extensions seem to have trained users not to expect it at all, so they might miss it regardless.

> 
> > I've also never had a format command fail in an IDE with an error. Just tried in JSON, ts in VS code and they do format even with syntax errors. Same for c# in VS.
> 
> Maybe it's just me then, but that behavior seems totally undesirable...

I agree with you that it's not ideal, but it seems like we'd be the only ones doing something about it. Being all exceptional and such is very tempting of course, but we had a similar argument when we discussed hovers (making them kind of like peek definition) and chose consistency with other VS code extensions there. Why do that there and this here?

> 
> > That said, they do produce weird formatting in some cases when there are syntax errors, so we don't need to be perfect here. If the user hates the result they can always undo.
> 
> Technically true, but there are going to be cases where it's not immediately obvious what has been changed - if you're working with larger files, without source control etc. Reformatting when we don't have matching brackets has the potential to introduce some very subtle changes that could go unnoticed for a while.

This is on the user to verify the changes made by tools. If they are using source control, it's easy to do a diff.

Reply:
@shenglol I don't think this discussion should block the PR btw. We can do another PR once we close on it.

Reply:
@majastrz Gotcha. I can merge the PR then once you and @anthony-c-martin approves.

Regarding formatting with syntax errors, I think I agree with @anthony-c-martin on the point that it is undesirable. Although there are cases it may work (like missing a comma in a JSON file), I find it confusing and even misleading a lot of times. Taking the VSCode JSON formatter as an example:

![code-formatting](https://user-images.githubusercontent.com/16367959/98291146-847beb80-1f5f-11eb-8801-91b2dba38b7e.gif)


The last formatting request with the missing `[` even changed the syntax error.

I'm trying to understand the use case where users would trigger formatting when there are syntax errors. One scenario might be that, say, I want to format a file, but I don't notice that I mistyped or deleted something. I cannot think of any good reasons why a user would format a file by intention when there are syntax errors. Usually when people notice syntax errors they will try to fix them immediately.

I think I'm still open to enable formatting broken syntax trees, but IMHO, we should do that only if it doesn't change the original syntax errors.

Reply:
Btw, I'm with both of you that it's kind of weird and potentially problematic to do it, but it seems like everyone else is doing it anyway. So if we do something different, we become the weird extension.

One case I can think of is formatting a file that's a work in progress. Your idea of skipping declarations containing syntax errors when formatting would take a care of it without causing weirdness in the unfinished parts of the file.